cluster-autoscaler/core/loop_state.go

```diff
@@ -0,0 +1,448 @@
+		ls.scaleDownForbidden = true
+	} else {
+		glog.V(4).Info("No schedulable pods")
+	}
+
+	return ls.onSuccess()
+}
+
+// See LoopStateHandler interface
+func (ls *loopState) ScaleUp() (ExecutionResult, errors.AutoscalerError) {
+	if len(ls.unschedulablePodsToHelp) == 0 {
+		glog.V(1).Info("No unschedulable pods")
+	} else if ls.autoscaler.MaxNodesTotal > 0 && len(ls.readyNodes) >= ls.autoscaler.MaxNodesTotal {
+		glog.V(1).Info("Max total nodes in cluster reached")
+	} else if allPodsAreNew(ls.unschedulablePodsToHelp, ls.currentTime) {
+		// The assumption here is that these pods have been created very recently and probably there
+		// is more pods to come. In theory we could check the newest pod time but then if pod were created
+		// slowly but at the pace of 1 every 2 seconds then no scale up would be triggered for long time.
+		// We also want to skip a real scale down (just like if the pods were handled).
+		ls.scaleDownForbidden = true
+		glog.V(1).Info("Unschedulable pods are very new, waiting one iteration for more")
+	} else {
+		daemonsets, err := ls.autoscaler.ListerRegistry.DaemonSetLister().List()
+		if err != nil {
+			glog.Errorf("Failed to get daemonset list")
+			return ls.onError(errors.ApiCallError, err)
+		}
+
+		scaleUpStart := time.Now()
+		metrics.UpdateLastTime(metrics.ScaleUp, scaleUpStart)
+
+		scaledUp, typedErr := ScaleUp(ls.context, ls.unschedulablePodsToHelp, ls.readyNodes, daemonsets)
+
+		metrics.UpdateDurationFromStart(metrics.ScaleUp, scaleUpStart)
+
+		if typedErr != nil {
+			glog.Errorf("Failed to scale up: %v", typedErr)
+			ls.onTypedError(typedErr)
+		} else if scaledUp {
+			ls.autoscaler.lastScaleUpTime = ls.currentTime
+			// No scale down in this iteration.
+			return ls.onEarlyExit()
+		}
+	}
+
+	return ls.onSuccess()
+}
+
+// See LoopStateHandler interface
+func (ls *loopState) ScaleDown() (ExecutionResult, errors.AutoscalerError) {
+	if !ls.autoscaler.ScaleDownEnabled {
+		return ls.onSuccess()
+	}
+
+	pdbs, err := ls.autoscaler.PodDisruptionBudgetLister().List()
+	if err != nil {
+		glog.Errorf("Failed to list pod disruption budgets: %v", err)
+		return ls.onError(errors.ApiCallError, err)
+	}
+
+	unneededStart := time.Now()
+
+	glog.V(4).Infof("Calculating unneeded nodes")
+
+	ls.autoscaler.scaleDown.CleanUp(ls.currentTime)
+	potentiallyUnneeded := getPotentiallyUnneededNodes(ls.context, ls.allNodes)
+
+	typedErr := ls.autoscaler.scaleDown.UpdateUnneededNodes(
+		ls.allNodes, potentiallyUnneeded,
+		append(ls.allScheduled, ls.unschedulableWaitingForLowerPriorityPreemption...),
+		ls.currentTime, pdbs)
+	if typedErr != nil {
+		glog.Errorf("Failed to scale down: %v", typedErr)
+		ls.onTypedError(typedErr)
+	}
+
+	metrics.UpdateDurationFromStart(metrics.FindUnneeded, unneededStart)
+
+	if glog.V(4) {
+		for key, val := range ls.autoscaler.scaleDown.unneededNodes {
+			glog.V(4).Infof("%s is unneeded since %s duration %s", key, val.String(), ls.currentTime.Sub(val).String())
```

Comment:
V(4) is checked twice - can we either remove it here, or delete the enclosing if?

Reply:
Thanks, I should have spotted this when moving the code (as I did move the if outside the loop).
Fixed.